python_modules/libraries/dagster-components/dagster_components/core/component.py

```diff
@@ -206,35 +213,76 @@ def get_registered_components_in_module(module: ModuleType) -> Iterable[Type[Com
 T = TypeVar("T")
 
 
+@record_custom
+class ComponentInstanceKey(IHaveNew):
+    """Uniquely identifies a component instance within a component hierarchy. Parts
+    typically correspond to the folder structure of the component hierarchy.
+    """
+
+    def __new__(cls, parts: Iterable[str]):
+        for part in parts:
+            if not is_valid_name(part):
+                check.param_invariant(False, "parts", f"Invalid key part {part}")
+
+        return super().__new__(cls, parts=parts)
```

Comment:
Note I tried a different approach using `NewType` to avoid duplicative runtime checking but `NewType` and `record` do not play nice together.